mobile_verifier/src/data_session.rs

```diff
@@ -0,0 +1,243 @@
+        tokio::spawn(async move {
+            loop {
+                tokio::select! {
+                    _ = shutdown.clone() => {
+                        tracing::info!("DataSessionIngestor shutting down");
+                        break;
+                    }
+                    Some(file) = receiver.recv() => self.process_file(file).await?,
+                }
+            }
+
+            Ok(())
+        })
+        .map_err(anyhow::Error::from)
+        .and_then(|result| async move { result })
+        .await
+    }
+
+    async fn process_file(
+        &self,
+        file_info_stream: FileInfoStream<ValidDataTransferSession>,
+    ) -> anyhow::Result<()> {
+        tracing::info!(
+            "handling valid data transfer file {:?}",
+            file_info_stream.file_info.key
+        );
+        let mut transaction = self.pool.begin().await?;
+        let file_ts = file_info_stream.file_info.timestamp;
+        file_info_stream
+            .into_stream(&mut transaction)
+            .await?
+            .map(anyhow::Ok)
+            .try_fold(transaction, |mut transaction, report| async move {
+                // if we have a subscriber id, then we want to save a subscriber data session
+                // entry to the DB
+                if let Some(ref subscriber_id) = report.subscriber_id {
+                    let data_session = SubscriberDataSession::from_valid_data_session(
+                        &report,
+                        subscriber_id.clone(),
+                        file_ts,
+                    );
+                    data_session.save(&mut transaction).await?;
+                    metrics::increment_counter!(
+                        "oracles_mobile_verifier_ingest_subscriber_data_session"
+                    )
+                }
+                // we always add a hotspot data session event
+                let data_session = HotspotDataSession::from_valid_data_session(report, file_ts);
+                data_session.save(&mut transaction).await?;
+                metrics::increment_counter!("oracles_mobile_verifier_ingest_hotspot_data_session");
+                Ok(transaction)
+            })
+            .await?
+            .commit()
+            .await?;
+        Ok(())
+    }
+}
+
+#[derive(sqlx::FromRow)]
+pub struct SubscriberDataSession {
+    pub subscriber_id: Vec<u8>,
+    pub upload_bytes: i64,
+    pub download_bytes: i64,
+    pub reward_timestamp: DateTime<Utc>,
+}
+
+#[derive(sqlx::FromRow)]
+pub struct HotspotDataSession {
+    pub pub_key: PublicKeyBinary,
+    pub payer: PublicKeyBinary,
+    pub upload_bytes: i64,
+    pub download_bytes: i64,
+    pub num_dcs: i64,
+    pub reward_timestamp: DateTime<Utc>,
+}
+
+#[derive(thiserror::Error, Debug)]
+#[error("data session error: {0}")]
+pub struct DataSessionError(#[from] sqlx::Error);
+
+impl SubscriberDataSession {
+    pub async fn save(self, db: &mut Transaction<'_, Postgres>) -> Result<(), DataSessionError> {
```

Comment:
in this case i'd skip the `DataSessionError` and just have the function return `Result<(), sqlx::Error>` since that's the only error this particular function can throw. the function calling this is returning an `anyhow::Result` which can safely handle receiving and throwing either the DataSessionError or the underlying sqlx::Error